### ISSUE TITLE
Issue while running on OpenJDK/OpenJFX

### ISSUE THREAD

**Issue by CrazyPyro** [RohitAwate/Everest#13]
Problem starting the application on a locked down Windows 10 Machine with JRE 8.
This may or may not be unique to my computer...will try on some others.

```
>java -version
java version "1.8.0_172"
Java(TM) SE Runtime Environment (build 1.8.0_172-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.172-b11, mixed mode)

> java -jar Everest-Alpha-1.0.jar
Failed to load native library:sqlite-3.21.0.1-5ca2c46c-1705-499a-a4b8-b7110fe961a0-sqlitejdbc.dll. osinfo: Windows/x86_64
java.lang.UnsatisfiedLinkError: C:\Users\...\AppData\Local\Temp\sqlite-3.21.0.1-5ca2c46c-1705-499a-a4b8-b7110fe961a0-sqlitejdbc.dll: This program is blocked by group policy. For more information, contact your system administrator
Connected to database.
Exception in thread "Thread-4" java.lang.UnsatisfiedLinkError: org.sqlite.core.NativeDB._open_utf8([BI)V
        at org.sqlite.core.NativeDB._open_utf8(Native Method)
        at org.sqlite.core.NativeDB._open(NativeDB.java:71)
        at org.sqlite.core.DB.open(DB.java:174)
        at org.sqlite.core.CoreConnection.open(CoreConnection.java:220)
        at org.sqlite.core.CoreConnection.<init>(CoreConnection.java:76)
        at org.sqlite.jdbc3.JDBC3Connection.<init>(JDBC3Connection.java:25)
        at org.sqlite.jdbc4.JDBC4Connection.<init>(JDBC4Connection.java:24)
        at org.sqlite.SQLiteConnection.<init>(SQLiteConnection.java:45)
        at org.sqlite.JDBC.createConnection(JDBC.java:114)
        at org.sqlite.JDBC.connect(JDBC.java:88)
        at java.sql.DriverManager.getConnection(Unknown Source)
        at java.sql.DriverManager.getConnection(Unknown Source)
        at com.rohitawate.everest.util.history.HistoryManager.<init>(HistoryManager.java:51)
        at com.rohitawate.everest.util.Services.lambda$start$0(Services.java:39)
        at java.lang.Thread.run(Unknown Source)
Settings file found. Loading settings...
Exception in thread "Settings loader thread" java.lang.NullPointerException
        at com.rohitawate.everest.util.logging.LoggingService.logInfo(LoggingService.java:52)
        at com.rohitawate.everest.util.settings.SettingsLoader.run(SettingsLoader.java:64)
        at java.lang.Thread.run(Unknown Source)
Exception in Application start method
java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at com.sun.javafx.application.LauncherImpl.launchApplicationWithArgs(LauncherImpl.java:389)
        at com.sun.javafx.application.LauncherImpl.launchApplication(LauncherImpl.java:328)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at sun.launcher.LauncherHelper$FXHelper.main(Unknown Source)
Caused by: java.lang.RuntimeException: Exception in Application start method
        at com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:917)
        at com.sun.javafx.application.LauncherImpl.lambda$launchApplication$154(LauncherImpl.java:182)
        at java.lang.Thread.run(Unknown Source)
Caused by: javafx.fxml.LoadException:
file:/.../Everest-Alpha-1.0.jar!/fxml/homewindow/HomeWindow.fxml
        at javafx.fxml.FXMLLoader.constructLoadException(FXMLLoader.java:2601)
        at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2579)
        at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2441)
        at javafx.fxml.FXMLLoader.load(FXMLLoader.java:2409)
        at com.rohitawate.everest.main.Main.start(Main.java:39)
        at com.sun.javafx.application.LauncherImpl.lambda$launchApplication1$161(LauncherImpl.java:863)
        at com.sun.javafx.application.PlatformImpl.lambda$runAndWait$174(PlatformImpl.java:326)
        at com.sun.javafx.application.PlatformImpl.lambda$null$172(PlatformImpl.java:295)
        at java.security.AccessController.doPrivileged(Native Method)
        at com.sun.javafx.application.PlatformImpl.lambda$runLater$173(PlatformImpl.java:294)
        at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:95)
        at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
        at com.sun.glass.ui.win.WinApplication.lambda$null$147(WinApplication.java:177)
        ... 1 more
Caused by: java.lang.NullPointerException
        at com.rohitawate.everest.util.logging.LoggingService.logInfo(LoggingService.java:52)
        at com.rohitawate.everest.controllers.HomeWindowController.recoverState(HomeWindowController.java:293)
        at com.rohitawate.everest.controllers.HomeWindowController.initialize(HomeWindowController.java:72)
        at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2548)
        ... 12 more
Exception running application com.rohitawate.everest.main.Main
```

**Comment by CrazyPyro** [comment]
On Fedora 28:
```
$ sudo dnf install java-1.8.0-openjdk
$ java -version
openjdk version "1.8.0_171"
OpenJDK Runtime Environment (build 1.8.0_171-b10)
OpenJDK 64-Bit Server VM (build 25.171-b10, mixed mode)

$ java -jar Everest-Alpha-1.0.jar 
Error: Could not find or load main class com.rohitawate.everest.main.Main
```

**Comment by CrazyPyro** [comment]
Update on Fedora 28 (Still doesn't work, but I had forgotten to install openjfx)  This might be better as a separate issue.
```
$ sudo dnf install openjfx
$ java -jar Everest-Alpha-1.0.jar 
Graphics Device initialization failed for :  es2, sw
Error initializing QuantumRenderer: no suitable pipeline found
java.lang.RuntimeException: java.lang.RuntimeException: Error initializing QuantumRenderer: no suitable pipeline found
        at com.sun.javafx.tk.quantum.QuantumRenderer.getInstance(QuantumRenderer.java:280)
        at com.sun.javafx.tk.quantum.QuantumToolkit.init(QuantumToolkit.java:221)
        at com.sun.javafx.tk.Toolkit.getToolkit(Toolkit.java:205)
        at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:209)
        at com.sun.javafx.application.LauncherImpl.startToolkit(LauncherImpl.java:675)
        at com.sun.javafx.application.LauncherImpl.launchApplicationWithArgs(LauncherImpl.java:337)
        at com.sun.javafx.application.LauncherImpl.launchApplication(LauncherImpl.java:328)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at sun.launcher.LauncherHelper$FXHelper.main(LauncherHelper.java:767)
Caused by: java.lang.RuntimeException: Error initializing QuantumRenderer: no suitable pipeline found
        at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.init(QuantumRenderer.java:94)
        at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.run(QuantumRenderer.java:124)
        at java.lang.Thread.run(Thread.java:748)
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at sun.launcher.LauncherHelper$FXHelper.main(LauncherHelper.java:767)
Caused by: java.lang.RuntimeException: No toolkit found
        at com.sun.javafx.tk.Toolkit.getToolkit(Toolkit.java:217)
        at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:209)
        at com.sun.javafx.application.LauncherImpl.startToolkit(LauncherImpl.java:675)
        at com.sun.javafx.application.LauncherImpl.launchApplicationWithArgs(LauncherImpl.java:337)
        at com.sun.javafx.application.LauncherImpl.launchApplication(LauncherImpl.java:328)
        ... 5 more
```

**Comment by RohitAwate** [comment]
`This program is blocked by group policy. For more information, contact your system administrator`
The Windows issue is definitely an isolated one since the machine is locked down and I'm afraid I can't do anything about it.

The Fedora one is a dependency issue. I could replicate it on Fedora 27.
Instead of `sudo yum install openjfx`, try `sudo dnf install java-1.8.0-openjdk-openjfx`. I would suggest uninstalling OpenJFX before doing this.

![screenshot from 2018-05-09 21-56-50](https://user-images.githubusercontent.com/23148259/39827144-b1a8278a-53d4-11e8-8d65-15d97be1b99a.png)

**Comment by CrazyPyro** [comment]
Thanks, I will test the OpenJFX thing on Fedora as soon as I get home.

On the work Windows machine, I know loading DLLs is a problem, and I will submit a patch to work around the Sqlite thing, but it looks like the FXML loading might be a separate issue.  Note that after the DLL error, it goes on with life and tries to load HomeWindow.fxml and that is where the final fatal error is.

I am not able to install [Oracle Scene Builder](http://www.oracle.com/technetwork/java/javase/downloads/javafxscenebuilder-1x-archive-2199384.html#javafx-scenebuilder-2.0-oth-JPR) but I tried to open that same file in [Gluon Scene Builder](http://gluonhq.com/products/scene-builder/) for Java 8, and got a very similar error complaining about the validity of the file itself.
![homewindow_fxml](https://user-images.githubusercontent.com/469161/39828987-fee92e8a-538a-11e8-903f-c3a326e2ce59.png)
Specifically, `ClassNotFoundException: com.jfoenix.controls.JFXButton` probably triggered by line 19 of the FXML: `<?import com.jfoenix.controls.JFXButton?>`
The underlying problem might involve https://github.com/jfoenixadmin/JFoenix/issues/134
I'll continue investigating tonight.  Thanks again for the help.
```
java.io.IOException: javafx.fxml.LoadException: 
/C:/Everest/src/main/resources/fxml/homewindow/HomeWindow.fxml

	at com.oracle.javafx.scenebuilder.kit.fxom.FXOMLoader.load(FXOMLoader.java:98)
	at com.oracle.javafx.scenebuilder.kit.fxom.FXOMDocument.<init>(FXOMDocument.java:85)
	at com.oracle.javafx.scenebuilder.kit.fxom.FXOMDocument.<init>(FXOMDocument.java:102)
	at com.oracle.javafx.scenebuilder.kit.editor.EditorController.updateFxomDocument(EditorController.java:2540)
	at com.oracle.javafx.scenebuilder.kit.editor.EditorController.setFxmlTextAndLocation(EditorController.java:761)
	at com.oracle.javafx.scenebuilder.app.DocumentWindowController.loadFromFile(DocumentWindowController.java:385)
	at com.oracle.javafx.scenebuilder.app.SceneBuilderApp.performOpenFiles(SceneBuilderApp.java:665)
	at com.oracle.javafx.scenebuilder.app.SceneBuilderApp.performOpenFile(SceneBuilderApp.java:601)
	at com.oracle.javafx.scenebuilder.app.SceneBuilderApp.performControlAction(SceneBuilderApp.java:204)
	at com.oracle.javafx.scenebuilder.app.welcomedialog.WelcomeDialogWindowController.openDocument(WelcomeDialogWindowController.java:153)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at sun.reflect.misc.Trampoline.invoke(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor1.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at sun.reflect.misc.MethodUtil.invoke(Unknown Source)
	at javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1771)
	at javafx.fxml.FXMLLoader$ControllerMethodEventHandler.handle(FXMLLoader.java:1657)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Node.fireEvent(Node.java:8411)
	at javafx.scene.control.ToggleButton.fire(ToggleButton.java:256)
	at com.sun.javafx.scene.control.behavior.ButtonBehavior.mouseReleased(ButtonBehavior.java:182)
	at com.sun.javafx.scene.control.skin.BehaviorSkinBase$1.handle(BehaviorSkinBase.java:96)
	at com.sun.javafx.scene.control.skin.BehaviorSkinBase$1.handle(BehaviorSkinBase.java:89)
	at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene$MouseHandler.process(Scene.java:3757)
	at javafx.scene.Scene$MouseHandler.access$1500(Scene.java:3485)
	at javafx.scene.Scene.impl_processMouseEvent(Scene.java:1762)
	at javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2494)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:394)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:295)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$353(GlassViewEventHandler.java:432)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:431)
	at com.sun.glass.ui.View.handleMouseEvent(View.java:555)
	at com.sun.glass.ui.View.notifyMouse(View.java:937)
	at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at com.sun.glass.ui.win.WinApplication.lambda$null$147(WinApplication.java:177)
	at java.lang.Thread.run(Unknown Source)
Caused by: javafx.fxml.LoadException: 
/C:/Everest/src/main/resources/fxml/homewindow/HomeWindow.fxml

	at javafx.fxml.FXMLLoader.constructLoadException(FXMLLoader.java:2601)
	at javafx.fxml.FXMLLoader.importClass(FXMLLoader.java:2848)
	at javafx.fxml.FXMLLoader.processImport(FXMLLoader.java:2692)
	at javafx.fxml.FXMLLoader.processProcessingInstruction(FXMLLoader.java:2661)
	at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2517)
	at javafx.fxml.FXMLLoader.load(FXMLLoader.java:2425)
	at com.oracle.javafx.scenebuilder.kit.fxom.FXOMLoader.load(FXOMLoader.java:93)
	... 67 more
Caused by: java.lang.ClassNotFoundException: com.jfoenix.controls.JFXButton
	at java.lang.ClassLoader.findClass(Unknown Source)
	at java.lang.ClassLoader.loadClass(Unknown Source)
	at java.lang.ClassLoader.loadClass(Unknown Source)
	at javafx.fxml.FXMLLoader.loadTypeForPackage(FXMLLoader.java:2916)
	at javafx.fxml.FXMLLoader.loadType(FXMLLoader.java:2905)
	at javafx.fxml.FXMLLoader.importClass(FXMLLoader.java:2846)
	... 72 more
```

**Comment by RohitAwate** [comment]
Hello again!
You'll need to import [JFoenix](http://www.jfoenix.com/download/jfoenix-8.0.4.jar) into Scene Builder before you open that FXML.

The ClassNotFoundException is quite surprising to me. I am not at all sure, but I still think it has got something to do with being run on your locked-down work machine.

I don't think I'll be able to reproduce this. Waiting to hear back from you.
Thanks! 👍

**Comment by RohitAwate** [comment]
That's great! Thank you for investing time into this, it may help someone in the future! :+1: 

Since you had to practically disable the logger, I'm curious, were there any files created in the `Everest\` directory? Is this some sort of a permissions issue? Maybe you're getting that `NullPointerException` because it couldn't write to the logs file?

**Comment by CrazyPyro** [comment]
3 files get created in .\Everest:

1. BugReporter.jar
2. config\everest.state   (some binary content)
3. logs\2018-05-10.html  (But no actual log messages; file ends with `<body><h1>Everest Log: 10/05/2018</h1><!-- Placeholder for new log --></body></html>`)

So it's not a file permissions thing.  Maybe something you did with the Services and threads and Executor gets a janky implementation on Windows. (Maybe consider something like log4j eventually?)

Anyway, thanks for making this.  I heard about you on a Hacker News article and this (specifically Java-based) solution was exactly what I needed!

**Comment by RohitAwate** [comment]
I see. I recently shifted to Google's Guava libraries for the Executor stuff. If possible, could you try out [this](https://github.com/RohitAwate/Everest/files/1990652/Everest-Alpha-1.0.zip) older build which uses plain Java Threads. I want to check if Guava is causing this issue.

I did look into log4j and it is really good. However, I decided to write a custom logger only because I couldn't get an HTML logger to work with log4j and I really wanted to write it myself just because it would be fun and interesting.

And finally, you're welcome! I'm glad you liked Everest.
Stay tuned for future updates! :)
